REF: improve extract method dialog UI

Kobzol opened the pull request:
This PR attempts to improve the Extract method dialog UI.
Related issues: https://github.com/intellij-rust/intellij-rust/issues/2186, https://github.com/intellij-rust/intellij-rust/issues/2183

Right now it fixes an exception that was thrown when an invalid name was used for the name of the function (https://github.com/intellij-rust/intellij-rust/issues/1934).

I also tried to use a simple text field instead of the dropdown list if there are no suggestions to offer, however `NameSuggestionsField` always returns true from `hasSuggestions()` method (which is caused by the way it's used in the dialog), so I'm not sure how to recognize when there are no suggestions (maybe use a different component altogether)?

ortem commented on src/main/kotlin/org/rust/ide/refactoring/extractFunction/ui.kt:
```diff
@@ -47,12 +53,50 @@ interface ExtractFunctionUi {
+
     override fun extract(config: RsExtractFunctionConfig, callback: () -> Unit) {
-        val functionNameField = NameSuggestionsField(project)
+        val functionNameField = this.createField()
```
Looks like our `Extract function` doesn't support name suggestions at all.

So it can be done this way, without `ExtractTextField`:
```Kotlin
val functionNameField = NameSuggestionsField(emptyArray(), project, RsFileType)
functionNameField.minimumSize = JBUI.size(300, 30)
```
`NameSuggestionsField` creates `EditorTextField` in case of empty suggestions array. Later we can pass real suggestions instead of `emptyArray()`.

Kobzol commented:
Okay, I removed that monstrosity and replaced it with an empty `NameSuggestionsField`. For some reason, even though I return a `ValidationInfo` with the name field, it is not focused after an error and the error is not shown until the user hovers upon the field.

ortem commented:
@Kobzol maybe just disable OK button (as Kotlin plugin does)? Something like
```Kotlin
val dialog = dialog(...) // without checkValidationErrors call
dialog.isOKActionEnabled = false
...
functionNameField.addDataChangedListener {
    updateConfig(config, functionNameField, visibilityBox)
    signatureComponent.setSignature(config.signature)
    dialog.isOKActionEnabled = checkValidationErrors(functionNameField).isEmpty()
}
```
What do you think?

Kobzol commented:
Thanks, I was actually looking for something like this. This is much more responsive.

ortem commented:
@Kobzol  Looks good! Do you want to add manual argument selection support (from the related issue) in this PR, or I can merge it?

Kobzol commented:
I'd like to add the parameters too. Is the Kotlin IDE source for this available?

ortem commented:
Yes, sure https://github.com/JetBrains/kotlin
Just open it in IDEA and search for “extractFunction”

Kobzol commented:
I added a parameter table to enable changing name, type and order of parameters.

I'm using `RsPsiFactory(project).tryCreateType(value)?.type ?: return` to create a type from a string, which has some unfortunate consequence. It doesn't take types that are available in scope into account and it doesn't allow to use parameter type such as `mut u32`. There could also be some completion/hints for the type names (any hints on how can I add that to the `TypeColumn`)?

Maybe it's not a good idea to allow the user to change the type arbitratily, since if he changes it, it is possible that the code will no longer compile after the refactoring. In Kotlin it will only allow to broaden the type (subtype -> supertype). Maybe we could add something similar (for example let the user choose from implemented `Traits` for a struct and then add a new generic parameter for that Trait?).
Another option would be to make the type immutable, but add a combobox column for changing ownership of the type (select `T`/`&T`/`&mut T`/`mut T`).

I also added a check for parameter and function names starting with `'` (since that is a valid identifier according to the name validator, but obviously not a valid function nor parameter name).

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/extractFunction/ui.kt:
```diff
@@ -67,23 +70,30 @@ private class DialogExtractFunctionUi(
-        functionNameField.addDataChangedListener {
-            updateConfig(config, functionNameField, visibilityBox)
+
+        val validator = LanguageNamesValidation.INSTANCE.forLanguage(RsLanguage)
```
I used that before, but `isValidRustVariableIdentifier` allows names starting with `'`, and that will crash the refactoring, since it's not a valid name for parameters/functions.

ortem replied:
`isValidRustVariableIdentifier` does not allow such names (in contrast to `isIdentifier`)

Kobzol replied:
Ooh, you're right of course. Thank you :)

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/extractFunction/ExtractFunctionParameterTablePanel.kt:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    // TODO: support for types in scope
+    override fun setValue(item: ParameterDataHolder, value: String) {
```
Ok.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/extractFunction/ui.kt:
```diff
@@ -67,23 +70,30 @@ private class DialogExtractFunctionUi(
             "Extract Function",
             panel,
-            resizable = false,
+            resizable = true,
```
I used `CCFlags.grow` for the rows with the name and signature components. Please let me know if there's a better way.

ortem commented:
Looks good, thank you! I updated the corresponding issue #2186. Changing argument's type/mutability/ownership and argument selection can be implemented in future PR.

bors r+
